### PR TITLE
Commerce 2014 marco

### PR DESCRIPTION
**Pull request by alecsloan**

### CODE REVIEW

**Comment by marco-leo** [comment]
ci:test

**Comment by ethanbustad** [comment]
^ @alecsloan I think that's a real compile issue -- not seeing it in upstream, and from the logs it looks legit.

**Comment by alecsloan** [comment]
ci:test

**Comment by alecsloan** [comment]
ci:test

**Comment by marco-leo** [comment]
ci:test

**Comment by alecsloan** [comment]
ci:test

**Comment by marco-leo** [comment]
ci:test

**Comment by marco-leo** [comment]
@alecsloan  this looks lilke a real fail

**Comment by ethanbustad** [comment]
ci:test

File: commerce-payment-service/src/main/java/com/liferay/commerce/payment/service/impl/CommercePaymentMethodGroupRelLocalServiceImpl.java [hunk 1]
```diff
@@ -306,6 +322,78 @@ public CommercePaymentMethodGroupRel setActive(
...
+							cpCatalogEntry.getCPDefinitionId());
+
+					if (cpDefinition != null) {
+						cpDefinition.setSubscriptionEnabled(false);
+						cpDefinition.setStatus(WorkflowConstants.STATUS_DRAFT);
+
+						_cpDefinitionLocalService.updateCPDefinition(
+							cpDefinition);
+					}
+				}
```
Comment (ethanbustad) [inline commerce-payment-service/src/main/java/com/liferay/commerce/payment/service/impl/CommercePaymentMethodGroupRelLocalServiceImpl.java]: @alecsloan can you put all this into a private method, maybe something like `_checkRecurringCommercePaymentMethods`? As-is, it feels really weird to have such a huge block inside a method named `setActive` -- which seems like it'd be simple -- that isn't really setting "active" at all, just taking care of a side effect. I think putting it into a private method that's called if the conditions are right (`!active`) will make a lot more sense in the future.

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java [hunk 2]
```diff
@@ -148,6 +176,41 @@ public PortletURL getPortletURL() {
...
+
+			return 0;
+		}
+
+		SearchContext searchContext = new SearchContext();
+
+		Map<String, Serializable> attributes = new HashMap<>();
+
+		attributes.put(Field.STATUS, WorkflowConstants.STATUS_APPROVED);
+
```
Comment (ethanbustad) [inline commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: no need for this newline

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java [hunk 3]
```diff
@@ -148,6 +176,41 @@ public PortletURL getPortletURL() {
...
+		attributes.put(CPField.PUBLISHED, true);
+		attributes.put(CPField.SUBSCRIPTION_ENABLED, true);
+
+		long groupId = commercePaymentMethodGroupRel.getGroupId();
```
Comment (ethanbustad) [inline commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: no need for this variable -- just use the getter where it's needed

Comment (ethanbustad) [reply commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: and after that, the `attributes.put` can all be one block

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java [hunk 4]
```diff
@@ -175,6 +238,8 @@ else if (navigation.equals("inactive")) {
 		_searchContainer = new SearchContainer<>(
 			_renderRequest, getPortletURL(), null, emptyResultsMessage);
 
+		long groupId = themeDisplay.getScopeGroupId();
```
Comment (ethanbustad) [inline commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: why add this variable? I think it was better before

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java [hunk 5]
```diff
@@ -223,6 +290,41 @@ public String getSelectedScreenNavigationCategoryKey() {
...
+						_commerceSubscriptionEntryLocalService.
+							fetchCommerceSubscriptionEntryByCommerceOrderItemId(
+								commerceOrderItem.getCommerceOrderItemId());
+
+					if (commerceSubscriptionEntry != null) {
+						activeSubscriptions++;
+					}
+				}
+			}
+		}
```
Comment (ethanbustad) [inline commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: oof.... is there any cleaner way we can do this? This feels very.... brute-force. And could get really slow in a large system.

Comment (ethanbustad) [reply commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: if this is all for supplying the warning message with more context for the user, it may not be worth it.....

File: commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp [hunk 6]
```diff
@@ -22,8 +22,25 @@ CommercePaymentMethodGroupRelsDisplayContext commercePaymentMethodGroupRelsDispl
...
 CommercePaymentMethodGroupRel commercePaymentMethodGroupRel = (CommercePaymentMethodGroupRel)row.getObject();
+
+String engineKey = commercePaymentMethodGroupRel.getEngineKey();
+
+String modalId = "modal" + engineKey;
+
+int subscriptionEntryCount = commercePaymentMethodGroupRelsDisplayContext.getSubscriptionEntryCount(engineKey);
+int recurringCPDefinitionsCount = commercePaymentMethodGroupRelsDisplayContext.getRecurringCPDefinitionsCount(commercePaymentMethodGroupRel);
+
+boolean isLastRecurringPaymentMethod = commercePaymentMethodGroupRelsDisplayContext.isLastRecurringPaymentMethod(engineKey);
```
Comment (ethanbustad) [inline commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp]: variable should not start with "is" -- I guess just `lastRecurringPaymentMethod` works.

File: commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp [hunk 7]
```diff
@@ -36,37 +53,121 @@ CommercePaymentMethodGroupRel commercePaymentMethodGroupRel = (CommercePaymentMe
...
-		<portlet:actionURL name="editCommercePaymentMethodGroupRel" var="setActiveURL">
-			<portlet:param name="<%= Constants.CMD %>" value="setActive" />
-			<portlet:param name="redirect" value="<%= currentURL %>" />
-			<portlet:param name="commercePaymentMethodGroupRelId" value="<%= String.valueOf(commercePaymentMethodGroupRel.getCommercePaymentMethodGroupRelId()) %>" />
-			<portlet:param name="active" value="<%= String.valueOf(!commercePaymentMethodGroupRel.isActive()) %>" />
-			<portlet:param name="engineKey" value="<%= commercePaymentMethodGroupRel.getEngineKey() %>" />
-		</portlet:actionURL>
-
 		<liferay-ui:icon
+			id='<%= "togglePaymentMethod-" + modalId %>'
```
Comment (ethanbustad) [inline commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp]: IDs shouldn't be dashed -- they should be mixedCase if possible; if not, just concatenate the strings together without a dash.

File: commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp [hunk 8]
```diff
@@ -36,37 +53,121 @@ CommercePaymentMethodGroupRel commercePaymentMethodGroupRel = (CommercePaymentMe
...
+		<style>
+			.warning-modal .modal-header {
+				background-color: #feefef;
+				border-color: #f48989;
+				color: #da1414;
+			}
+
+			.warning-modal p {
+				font-size: 16px;
+			}
```
Comment (ethanbustad) [inline commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp]: this seems..... overly specific. If there's any way we can add more classes to the `div` to hook into existing styling, rather than adding this custom styling just for this case (which might easily clash with somebody's theme down the road), that would be much better.

File: commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp [hunk 9]
```diff
@@ -36,37 +53,121 @@ CommercePaymentMethodGroupRel commercePaymentMethodGroupRel = (CommercePaymentMe
...
+		<%
+			String deactivationWarning = StringPool.BLANK;
+
+			if (subscriptionEntryCount > 0) {
+				deactivationWarning += LanguageUtil.format(locale, "there-are-x-subscriptions-that-depend-on-this-payment-method-if-you-deactivate-it-you-will-have-to-manually-manage-those-subscriptions", subscriptionEntryCount);
+			}
+
+			if (isLastRecurringPaymentMethod) {
+				deactivationWarning += "<br /><br />" + LanguageUtil.format(locale, "deactivating-this-payment-method-will-disable-x-subcription-products", recurringCPDefinitionsCount);
+			}
```
Comment (ethanbustad) [inline commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp]: deindent one level

File: commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp [hunk 10]
```diff
@@ -36,37 +53,121 @@ CommercePaymentMethodGroupRel commercePaymentMethodGroupRel = (CommercePaymentMe
...
+		%>
+
+		<aui:script use="aui-base">
+			A.use(
+				'aui-modal',
+				function(A) {
+					var <%= modalId %> = new A.Modal(
+						{
+							bodyContent: '<p><liferay-ui:message key="<%= deactivationWarning %>" /><p>',
```
Comment (ethanbustad) [inline commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp]: we've already localized the warning by this point, so we should remove the `liferay-ui:message` and just put the warning in the `<p>`

File: commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp (hunk 9 above)
Comment (ethanbustad) [reply commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp]: also: if the first condition is false, but the second condition is true, the message will start with two newlines, which is a bit weird. Might be better to add the warnings to a list, and then `StringUtil.merge` the list with `<br /><br />` as the delimiter

File: commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp [hunk 11]
```diff
@@ -36,37 +53,121 @@ CommercePaymentMethodGroupRel commercePaymentMethodGroupRel = (CommercePaymentMe
...
+							centered: true,
+							draggable: false,
+							destroyOnHide: false,
+							headerContent: '<h2><liferay-ui:message key="are-you-sure" /></h2>',
+							modal: true,
+							boundingBox: '#<%= modalId %>',
+							width: 450
```
Comment (ethanbustad) [inline commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp]: sort, please :)

File: commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceSubscriptionEntryLocalServiceImpl.java [hunk 12]
```diff
@@ -280,8 +280,9 @@ public CommerceSubscriptionEntry fetchCommerceSubscriptionEntries(
 	}
 
 	@Override
-	public CommerceSubscriptionEntry fetchCommerceSubscriptionEntry(
-		long commerceOrderItemId) {
+	public CommerceSubscriptionEntry
+		fetchCommerceSubscriptionEntryByCommerceOrderItemId(
```
Comment (ethanbustad) [inline commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceSubscriptionEntryLocalServiceImpl.java:284]: 🙏

File: commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java [hunk 13]
```diff
@@ -85,6 +99,48 @@ public PortletURL getPortletURL() throws PortalException {
 		return portletURL;
 	}
 
+	public boolean hasRecurringPaymentMethod() throws PortalException {
+		List<CommerceChannelRel> commerceChannelRels =
+			CommerceChannelRelLocalServiceUtil.getCommerceChannelRels(
+				CPDefinition.class.getName(), getCPDefinitionId(), -1, -1,
```
Comment (ethanbustad) [inline commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java]: `QueryUtil.ALL_POS` rather than `-1`

File: commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java [hunk 14]
```diff
@@ -85,6 +99,48 @@ public PortletURL getPortletURL() throws PortalException {
...
+
+					if (commercePaymentMethod.isProcessRecurringEnabled()) {
+						channelHasRecurringPaymentMethod = true;
+					}
+				}
+			}
+
+			if (!channelHasRecurringPaymentMethod) {
+				return false;
+			}
```
Comment (ethanbustad) [inline commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java]: Is this the logic we want? This seems to be saying:
 > Get all channels that publish this product. If any of those channels doesn't have a recurring payment method, then for this product `hasRecurringPaymentMethod=false`

I think I would expect:
 > Get all channels that publish this product. If any of those channels has a recurring payment method, then for this product `hasRecurringPaymentMethod=true`
Can you confirm that it's correct as-is?

File: commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java [hunk 15]
```diff
@@ -85,6 +99,48 @@ public PortletURL getPortletURL() throws PortalException {
...
+					getCommercePaymentMethodGroupRels(
+						commerceChannel.getSiteGroupId(), true);
+
+			for (CommercePaymentMethodGroupRel commercePaymentMethodGroupRel :
+					commercePaymentMethodGroupRels) {
+
+				if (commercePaymentMethodGroupRel.isActive()) {
+					CommercePaymentMethod commercePaymentMethod =
+						_commercePaymentMethodRegistry.getCommercePaymentMethod(
+							commercePaymentMethodGroupRel.getEngineKey());
```
Comment (ethanbustad) [inline commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java]: also, this again seems a bit problematic to me. Doing IO inside a nested loop is a recipe for performance issues....

File: commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPInstanceSubscriptionInfoDisplayContext.java [hunk 16]
```diff
@@ -94,6 +116,50 @@ public PortletURL getPortletURL() throws PortalException {
...
+
+					if (commercePaymentMethod.isProcessRecurringEnabled()) {
+						channelHasRecurringPaymentMethod = true;
+					}
+				}
+			}
+
+			if (!channelHasRecurringPaymentMethod) {
+				return false;
+			}
```
Comment (ethanbustad) [inline commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPInstanceSubscriptionInfoDisplayContext.java]: same questions/concerns here

File: commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java (hunk 14 above)
Comment (ethanbustad) [reply commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java]: Oh interesting, just saw:
> all-channels-associated-with-this-product-must-have-at-least-one-payment-method-active-that-supports-recurring-payments

Guess that is what we want :)

File: commerce-subscription-web/src/main/resources/META-INF/resources/subscription_entry_action.jsp [hunk 17]
```diff
@@ -22,6 +22,14 @@ CommerceSubscriptionEntryDisplayContext commerceSubscriptionEntryDisplayContext
 ResultRow row = (ResultRow)request.getAttribute(WebKeys.SEARCH_CONTAINER_RESULT_ROW);
 
 CommerceSubscriptionEntry commerceSubscriptionEntry = (CommerceSubscriptionEntry)row.getObject();
+
+int subscriptionStatus = CommerceSubscriptionEntryConstants.SUBSCRIPTION_STATUS_ANY;
```
Comment (ethanbustad) [inline commerce-subscription-web/src/main/resources/META-INF/resources/subscription_entry_action.jsp:24]: should be sorted above `CommerceSubscriptionEntry commerceSubscriptionEntry`

File: commerce-subscription-web/src/main/resources/META-INF/resources/view.jsp [hunk 18]
```diff
@@ -164,6 +166,19 @@ boolean hasManageCommerceSubscriptionEntryPermission = commerceSubscriptionEntry
 						value="<%= String.valueOf(commerceOrder.getCommerceAccountId()) %>"
 					/>
 
+					<%
+					String paymentMethod = LanguageUtil.get(request, commerceOrder.getCommercePaymentMethodKey());
+
+					if (!commerceSubscriptionEntryDisplayContext.isPaymentMethodActive(commerceOrder.getCommercePaymentMethodKey())) {
+						paymentMethod += StringPool.DOUBLE_SPACE + StringPool.OPEN_PARENTHESIS + LanguageUtil.get(request, "inactive") + StringPool.CLOSE_PARENTHESIS;
```
Comment (ethanbustad) [inline commerce-subscription-web/src/main/resources/META-INF/resources/view.jsp]: we should have a language key for this. concatenating stuff to a localized value is always wrong. (or at least almost always, but I can't think of an exception atm.) the key can just be like `x-inactive`, and we can pass the other localized value in.

File: commerce-subscription-web/src/main/resources/META-INF/resources/definition/edit_definition_subscription_info.jsp [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+
+CPSubscriptionType cpSubscriptionType = cpDefinitionSubscriptionInfoDisplayContext.getCPSubscriptionType(subscriptionType);
+
+if (cpSubscriptionType != null) {
+	defaultCPSubscriptionTypeLabel = cpSubscriptionType.getLabel(locale);
+}
+
+CPSubscriptionTypeJSPContributor cpSubscriptionTypeJSPContributor = cpDefinitionSubscriptionInfoDisplayContext.getCPSubscriptionTypeJSPContributor(subscriptionType);
+
+boolean ending = maxSubscriptionCycles > 0;
```
Comment (ethanbustad) [inline commerce-subscription-web/src/main/resources/META-INF/resources/definition/edit_definition_subscription_info.jsp]: ugh, looks like this file may have just been moved... but git shows it as deleted + added. Oh well.

Don't set a boolean to the result of an operation like this. use `ending = false; if (blah) ending = true;` (with better formatting obviously)

**Review by ethanbustad — CHANGES_REQUESTED** [review]

**Comment by ethanbustad** [comment]
@marco-leo obviously I've requested a few changes here, but nothing is a deal-breaker for release -- so if you want to merge this now to get the release finalized, and we make these improvements afterward, that's ok with me.

**Comment by marco-leo** [comment]
@ethanbustad yeap! sounds good to me, then I will review one more time as is a big pull.

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java (hunk 5 above)
Comment (ethanbustad) [reply commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: Yeah this strikes me as something we should use the indexer for -- maybe even just add payment method key to the subscription entry document, and then query for a count that way.

File: commerce-payment-service/src/main/java/com/liferay/commerce/payment/service/impl/CommercePaymentMethodGroupRelLocalServiceImpl.java [hunk 20]
```diff
@@ -306,6 +330,77 @@ public CommercePaymentMethodGroupRel setActive(
...
+					lastRecurringCommercePaymentMethod.getKey();
+
+				if (!lastRecurringCommercePaymentMethodKey.equals(
+						commercePaymentMethodGroupRel.getEngineKey())) {
+
+					return commercePaymentMethodGroupRelPersistence.update(
+						commercePaymentMethodGroupRel);
+				}
+
+				SearchContext searchContext = new SearchContext();
```
Comment (marco-leo) [inline commerce-payment-service/src/main/java/com/liferay/commerce/payment/service/impl/CommercePaymentMethodGroupRelLocalServiceImpl.java]: @alecsloan why we are using search framewrok and not a simpler service call ?

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java (hunk 3 above)
Comment (marco-leo) [reply commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: This looks a replication of the logic. And still I don't see value in using search over standard finders

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java [hunk 21]
```diff
@@ -223,6 +290,41 @@ public String getSelectedScreenNavigationCategoryKey() {
...
+		ThemeDisplay themeDisplay = (ThemeDisplay)_renderRequest.getAttribute(
+			WebKeys.THEME_DISPLAY);
+
+		int activeSubscriptions = 0;
+
+		long groupId =
+			_commerceChannelLocalService.getCommerceChannelGroupIdBySiteGroupId(
+				themeDisplay.getSiteGroupId());
+
+		List<CommerceOrder> commerceOrders =
```
Comment (marco-leo) [inline commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: Are we fetching all orders without pagination? Do we understand that this can be a very big number?

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java [hunk 22]
```diff
@@ -223,6 +290,41 @@ public String getSelectedScreenNavigationCategoryKey() {
...
+			_commerceOrderLocalService.getCommerceOrders(groupId, engineKey);
+
+		for (CommerceOrder commerceOrder : commerceOrders) {
```
Comment (marco-leo) [inline commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: for inside a for, I would always prefer to avoid this scenario, make code hard to follow

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java [hunk 23]
```diff
@@ -223,6 +290,41 @@ public String getSelectedScreenNavigationCategoryKey() {
...
+					commerceOrder.getCommerceOrderItems()) {
+
+				if (commerceOrderItem.isSubscription()) {
+					CommerceSubscriptionEntry commerceSubscriptionEntry =
+						_commerceSubscriptionEntryLocalService.
+							fetchCommerceSubscriptionEntryByCommerceOrderItemId(
+								commerceOrderItem.getCommerceOrderItemId());
+
+					if (commerceSubscriptionEntry != null) {
+						activeSubscriptions++;
```
Comment (marco-leo) [inline commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: creating a finder will simplify a lot this code and avoid all of this object instantiation

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java [hunk 24]
```diff
@@ -232,6 +334,33 @@ public boolean hasManageCommercePaymentMethodGroupRelsPermission() {
 			CommerceActionKeys.MANAGE_COMMERCE_PAYMENT_METHODS);
 	}
 
+	public boolean isLastRecurringPaymentMethod(String engineKey) {
```
Comment (marco-leo) [inline commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java:255]: why?

File: commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp [hunk 25]
```diff
@@ -36,37 +53,121 @@ CommercePaymentMethodGroupRel commercePaymentMethodGroupRel = (CommercePaymentMe
...
 			url="<%= editURL %>"
 		/>
 	</c:if>
-</liferay-ui:icon-menu>
\ No newline at end of file
+</liferay-ui:icon-menu>
+
+<c:choose>
+	<c:when test="<%= ((subscriptionEntryCount > 0) || isLastRecurringPaymentMethod) && commercePaymentMethodGroupRel.isActive() %>">
+		<style>
```
Comment (marco-leo) [inline commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp]: never put the style in jsp.  Add those css classes to the portlet css file and apply those calsses to divs

File: commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp [hunk 26]
```diff
@@ -36,37 +53,121 @@ CommercePaymentMethodGroupRel commercePaymentMethodGroupRel = (CommercePaymentMe
...
+			if (subscriptionEntryCount > 0) {
+				deactivationWarning += LanguageUtil.format(locale, "there-are-x-subscriptions-that-depend-on-this-payment-method-if-you-deactivate-it-you-will-have-to-manually-manage-those-subscriptions", subscriptionEntryCount);
+			}
+
+			if (isLastRecurringPaymentMethod) {
+				deactivationWarning += "<br /><br />" + LanguageUtil.format(locale, "deactivating-this-payment-method-will-disable-x-subcription-products", recurringCPDefinitionsCount);
+			}
+		%>
+
+		<aui:script use="aui-base">
```
Comment (marco-leo) [inline commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp:81]: is this the equivalent of a javascript confirm? In that case, seem to be a bit complext

File: commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java [hunk 27]
```diff
@@ -85,6 +99,48 @@ public PortletURL getPortletURL() throws PortalException {
 		return portletURL;
 	}
 
+	public boolean hasRecurringPaymentMethod() throws PortalException {
+		List<CommerceChannelRel> commerceChannelRels =
+			CommerceChannelRelLocalServiceUtil.getCommerceChannelRels(
+				CPDefinition.class.getName(), getCPDefinitionId(), -1, -1,
+				null);
+
+		for (CommerceChannelRel commerceChannelRel : commerceChannelRels) {
```
Comment (marco-leo) [inline commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java]: This code is really hard to maintain. I don't like it, let's find an easier implementation

File: commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPInstanceSubscriptionInfoDisplayContext.java [hunk 28]
```diff
@@ -94,6 +116,50 @@ public PortletURL getPortletURL() throws PortalException {
 		return portletURL;
 	}
 
+	public boolean hasRecurringPaymentMethod() throws PortalException {
+		List<CommerceChannelRel> commerceChannelRels =
+			_commerceChannelRelLocalService.getCommerceChannelRels(
+				CPDefinition.class.getName(), getCPDefinitionId(), -1, -1,
+				null);
+
+		for (CommerceChannelRel commerceChannelRel : commerceChannelRels) {
```
Comment (marco-leo) [inline commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPInstanceSubscriptionInfoDisplayContext.java]: seems to be a duplication, please respect DRY

**Review by marco-leo — CHANGES_REQUESTED** [review]

File: commerce-payment-service/src/main/java/com/liferay/commerce/payment/service/impl/CommercePaymentMethodGroupRelLocalServiceImpl.java (hunk 20 above)
Comment (alecsloan) [reply commerce-payment-service/src/main/java/com/liferay/commerce/payment/service/impl/CommercePaymentMethodGroupRelLocalServiceImpl.java]: I'll add a finder for getting definitions by subscription enabled and channelGroupId

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java (hunk 21 above)
Comment (alecsloan) [reply commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: All orders for the payment method they're disabling. I'll try to think of a better way for checking active subscriptions

Comment (alecsloan) [reply commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: Dropping this whole method. Only needed for a custom message so not that important.

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java (hunk 24 above)
Comment (alecsloan) [reply commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java:255]: If its the last recurring payment method we disable subscription products since there's no longer a payment method to purchase them with.

File: commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp (hunk 26 above)
Comment (alecsloan) [reply commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp]: how come complex? It's not just a regular javascript confirm. It's a multi-option modal.

Comment (alecsloan) [reply commerce-payment-web/src/main/resources/META-INF/resources/payment_method_action.jsp:81]: The point was to have a custom modal to indicate how serious the action was but I understand.

File: commerce-payment-service/src/main/java/com/liferay/commerce/payment/service/impl/CommercePaymentMethodGroupRelLocalServiceImpl.java (hunk 20 above)
Comment (alecsloan) [reply commerce-payment-service/src/main/java/com/liferay/commerce/payment/service/impl/CommercePaymentMethodGroupRelLocalServiceImpl.java]: After changing to a finder and testing I remembered that I was using the search because we need to look at all the products in the current channel. With the finder we would have to look in specific catalogs.

Comment (alecsloan) [reply commerce-payment-service/src/main/java/com/liferay/commerce/payment/service/impl/CommercePaymentMethodGroupRelLocalServiceImpl.java]: Removing that custom message also. Instead of telling the admin how many products it will disable, the message will just be general.

File: commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java (hunk 27 above)
Comment (alecsloan) [reply commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java]: As discussed in slack I'm not sure what other approach we can take

File: commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java (hunk 5 above)
Comment (alecsloan) [reply commerce-payment-web/src/main/java/com/liferay/commerce/payment/web/internal/display/context/CommercePaymentMethodGroupRelsDisplayContext.java]: Fixed with finders

File: commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java (hunk 15 above)
Comment (alecsloan) [reply commerce-subscription-web/src/main/java/com/liferay/commerce/subscription/web/internal/display/context/CPDefinitionSubscriptionInfoDisplayContext.java]: Yeah, I'll have to talk to @marco-leo about it when I talk to him about the other stuff

**Comment by ethanbustad** [comment]
ci:test

**Comment by alecsloan** [comment]
ci:test

**Comment by alecsloan** [comment]
ci:test

**Comment by alecsloan** [comment]
ci:test

File: commerce-lang/src/main/resources/content/Language.properties [hunk 29]
```diff
@@ -559,7 +559,7 @@ there-is-a-newer-version-of-this-product-available=There is a new version of thi
 there-is-already-one-sku-with-the-selected-options=There is already one SKU with the selected options.
 there-is-no-option-selected=There is no option selected.
 there-is-no-order-selected=There is no order selected.
-there-might-be-subscriptions-that-depend-on-this-payment-method-if-you-deactivate-it-you-will-have-to-manually-manage-those-subscriptions-and-all-subcription-products-in-the-current-channel-will-be-disabled=There might be subscriptions that depend on this payment method. If you deactivate it you will have to manually manage those subscriptions and all subscription products in the current channel will be disabled.
+there-might-be-subscriptions-that-depend-on-this-payment-method-if-you-deactivate-it-you-will-have-to-manually-manage-those-subscriptions=There might be subscriptions that depend on this payment method. If you deactivate it you will have to manually manage those subscription.
```
Comment (ethanbustad) [inline commerce-lang/src/main/resources/content/Language.properties:562]: `...those subscriptions.`, not `...those subscription.`
